plugins/ICL1ObjectProducer.cc

```diff
@@ -0,0 +1,200 @@
+        for (l1t::JetBxCollection::const_iterator it=jet->begin(0); it!=jet->end(0); it++){
+            ic::L1TJet thisJet;
+            ROOT::Math::PtEtaPhiEVector tempVector(it->pt(), it->eta(),it->phi(), it->energy());
+            thisJet.set_vector(tempVector);
+            L1Jets_->push_back(thisJet);
+        }
+    //}
+    
+  } else {
+    edm::LogWarning("MissingProduct") << "L1Upgrade Jets not found. Branch will not be filled" << std::endl;
+  }
+
+  if (sums.isValid()){ 
+    
+    //for (int ibx = sums->getFirstBX(); ibx <= sums->getLastBX(); ++ibx) {
+        for (l1t::EtSumBxCollection::const_iterator it=sums->begin(0); it!=sums->end(0); it++){
+            ic::L1TSum thisSum;
+            int type = static_cast<int>( it->getType() );
+            thisSum.sumType = type;
+            //std::cout << type << "    " << << std::endl;
+            ROOT::Math::PtEtaPhiEVector tempVector(it->pt(), it->eta(),it->phi(), it->energy());
+            thisSum.set_vector(tempVector);
+            L1Sums_->push_back(thisSum);
+        }
+    //}
+    
+  } else {
+    edm::LogWarning("MissingProduct") << "L1Upgrade EtSums not found. Branch will not be filled" << std::endl;
+  }
+
+  if (muon.isValid()){ 
+    
+    //for (int ibx = muon->getFirstBX(); ibx <= muon->getLastBX(); ++ibx) {
+        for (l1t::MuonBxCollection::const_iterator it=muon->begin(0); it!=muon->end(0); it++){
+            ic::L1TMuon thisMuon;
+            thisMuon.isolation = it->hwIso();
+            thisMuon.charge = it->charge(); 
+            thisMuon.quality = it->hwQual();
+            ROOT::Math::PtEtaPhiEVector tempVector(it->pt(), it->eta(),it->phi(), it->energy());
+            thisMuon.set_vector(tempVector);
+            L1Muons_->push_back(thisMuon);
+        }
+    //}
+    
+    
+  } else {
+    edm::LogWarning("MissingProduct") << "L1Upgrade Muons not found. Branch will not be filled" << std::endl;
+  }
+
+  if (tau.isValid()){ 
+    
+    //for (int ibx = tau->getFirstBX(); ibx <= tau->getLastBX(); ++ibx) {
+        for (l1t::TauBxCollection::const_iterator it=tau->begin(0); it!=tau->end(0); it++){
+            ic::L1TTau thisTau;
+            thisTau.isolation = it->hwIso();
+            ROOT::Math::PtEtaPhiEVector tempVector(it->pt(), it->eta(),it->phi(), it->energy());
+            thisTau.set_vector(tempVector);
+            L1Taus_->push_back(thisTau);
+        }
+    //}
+    
+
+  } else {
+    edm::LogWarning("MissingProduct") << "L1Upgrade Tau not found. Branch will not be filled" << std::endl;
+  }
+
+}
+
+
+void ICL1ObjectProducer::beginJob() {
+    
+  ic::StaticTree::tree_->Branch("L1Muons", &L1Muons_);
+  ic::StaticTree::tree_->Branch("L1Taus", &L1Taus_);
+  ic::StaticTree::tree_->Branch("L1EGammas", &L1EGammas_);
+  ic::StaticTree::tree_->Branch("L1Jets", &L1Jets_);
+  ic::StaticTree::tree_->Branch("L1Sums", &L1Sums_);
+  
+}
+
+void ICL1ObjectProducer::endJob() {}
+
+DEFINE_FWK_MODULE(ICL1ObjectProducer);
```

Comment:
Hi, I agree, there are a few things that it would be nice to improve.
 
 - A trivial one, but the header includes at the top aren't directly including the L1Candidate data formats that we're reading, instead they're being picked up via some other random L1ntuples package. We should have direct includes to DataFormats/L1Trigger/interface/.
 - I'm generally not a fan of producing multiple collections in one producer, in the long run it's less flexible - what if in the future there are two L1Tau collections we need to read for example?
 - The CMSSW [base class](https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/DataFormats/L1Trigger/interface/L1Candidate.h) defines a set of [common integers](https://github.com/cms-sw/cmssw/blob/CMSSW_8_0_X/DataFormats/L1Trigger/interface/L1Candidate.h#L48-L61) that all the derived classes carry. In our own L1 data formats this already covers almost all the variables we seem to want to save: ic::L1TEGamma gets hwIso(), ic::L1TJet doesn't get anything specific, ic::L1TMuon gets hwIso(), hwQual() and charge() (but ic::Candidate can already store the latter, so we definitely shouldn't be adding a new charge data member like we do now), and ic::L1TTau gets hwIso(). The one exception appears to be the ic::L1TSum, which needs a type integer. However in the ic::Candidate base class you could just recycle id() or charge() to store this - I don't think it's worth writing a whole new class. 
 - Would also avoid setting data members directly - in the rest of the framework these are private and we use getters and setters. It's more future-proof to keep the interface separate, even for what are currently simple variables.
 - So I'd suggest having one ic::L1Candidate class deriving directly from ic::Candidate that has isolation (from hwIso) and quality (from hwQual) data members. Then we only need one templated CMSSW producer that will produce one collection, and that fills the p4, charge and these two variables in the default implementation, and once specialisation of the produce method for the L1TSum to get the type. Could even delegate to constructSpecific function like we do in ICTauProducer for example.